compiler/rustc_query_system/src/query/mod.rs

```diff
@@ -93,10 +93,12 @@ pub struct QuerySideEffects {
 }
 
 impl QuerySideEffects {
+    /// Returns true if there might be side effects.
     #[inline]
-    pub fn is_empty(&self) -> bool {
-        let QuerySideEffects { diagnostics } = self;
-        diagnostics.is_empty()
+    pub fn maybe_any(&self) -> bool {
+        // Use `has_capacity` so that the destructor for `self.diagnostics` can be skipped
+        // if `maybe_any` is known to be false.
+        self.diagnostics.has_capacity()
```

Comment:
Could you keep the destructuring assignment? That's a good way to ensure we don't forget anything in a refactor.